File: python/demo_fed_main.py
#!/usr/bin/env python3
"""Minimal FED demonstration (main release, Python)."""

from __future__ import annotations

import argparse
from pathlib import Path

import numpy as np

from fed_cov_main import fed_cov_main


def build_synthetic_signal(fs: int = 128, duration: int = 4, seed: int = 42) -> tuple[np.ndarray, np.ndarray]:
    rng = np.random.default_rng(seed)
    t_len = fs * duration
    t = np.arange(t_len) / fs

    c1 = 1.0 * np.sin(2 * np.pi * 3 * t)
    c2 = 0.7 * np.sin(2 * np.pi * 12 * t) * (0.6 + 0.5 * np.sin(2 * np.pi * 0.7 * t))
    c3 = np.zeros(t_len)
    for bt in (0.5, 1.5, 2.5, 3.5):
        c3 += 0.5 * np.exp(-((t - bt) ** 2) / (2 * 0.1**2)) * np.sin(2 * np.pi * 30 * t)

    x = c1 + c2 + c3 + 0.1 * rng.standard_normal(t_len)
    return t, x


def print_row(name: str, out: dict) -> None:
    print(
        f"{name:<22}  {out['epsilon_L2']:5.3f}   "
        f"{out['epsilon_F_local']:9.3f}   {out['F_selectivity']:5.3f}   "
        f"{out['Finv_orthogonality_deviation']:8.1e}"
    )


def main() -> None:
    parser = argparse.ArgumentParser(description="FED main demo (Python)")
    parser.add_argument("--fs", type=int, default=128)
    parser.add_argument("--duration", type=int, default=4)
    parser.add_argument("--L", type=int, default=48)
    parser.add_argument("--num-modes", type=int, default=12)
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--save-fig", action="store_true", default=True)
    parser.add_argument("--no-save-fig", dest="save_fig", action="store_false")
    parser.add_argument("--out-dir", type=Path, default=Path(__file__).resolve().parent / "output")
    args = parser.parse_args()

    t, x = build_synthetic_signal(fs=args.fs, duration=args.duration, seed=args.seed)
    n = x.size

    print("\n=== FED main demo (Python) ===")
    print(f"Fs={args.fs} Hz, N={n}, L={args.L}, k={args.num_modes}\n")

    common = {
        "L": args.L,
        "num_modes": args.num_modes,
        "f_strategy": "row",
        "kmax": 6,
        "verbose": True,
    }

    print("--- SSA baseline (F = I) ---")
    out_ssa = fed_cov_main(x, **common, f_transform="identity", f_param=1)

    print("\n--- FED default (F = HFD^2, p = 2) ---")
    out_fed = fed_cov_main(x, **common, f_transform="power", f_param=2)

    print("\n=== Summary ===")
    print(f"{'Method':<22}  epsL2   epsF_high   F-Sel   Finv_orth")
    print("-" * 58)
    print_row("SSA (F=I)", out_ssa)
    print_row("FED (HFD^2,p=2)", out_fed)

    if args.save_fig:
        try:
            import matplotlib.pyplot as plt
        except ImportError:
            print("\nmatplotlib not installed; skipping figure export.")
            print("Install with: pip install matplotlib")
        else:
            args.out_dir.mkdir(parents=True, exist_ok=True)
            fig, axes = plt.subplots(1, 2, figsize=(11, 4.2), facecolor="white")

            for ax, out, title in (
                (axes[0], out_ssa, f"SSA (F=I)  epsL2={out_ssa['epsilon_L2']:.3f}"),
                (axes[1], out_fed, f"FED (HFD^2)  epsL2={out_fed['epsilon_L2']:.3f}"),
            ):
                ax.plot(t, x, color=(0.2, 0.2, 0.2), label="Signal")
                ax.plot(t, out["x_hat"], linewidth=1.1, label="Reconstruction")
                ax.grid(True)
                ax.set_xlabel("Time (s)")
                ax.set_ylabel("Amplitude")
                ax.set_title(title)
                ax.legend(loc="best")

            fpath = args.out_dir / "demo_fed_main_reconstruction.png"
            fig.tight_layout()
            fig.savefig(fpath, dpi=150)
            plt.close(fig)
            print(f"\nFigure saved: {fpath}")

    print("\nDone. Run verify_fed_main.py for the Step-6 algebra check.")


if __name__ == "__main__":
    main()

File: matlab/demo_fed_main.m
function demo_fed_main(varargin)
%DEMO_FED_MAIN  Minimal FED demonstration (main release).
%
%   Builds the standard synthetic test signal, runs fed_cov_main for:
%     (1) F = I  (SSA baseline)
%     (2) F = HFD^2, p = 2  (paper default)
%   Prints reconstruction metrics and saves an optional figure.
%
%   Usage (from this folder):
%     demo_fed_main
%     demo_fed_main('SaveFig', false)
%
%   See also: fed_cov_main, verify_fed_main

here = fileparts(mfilename('fullpath'));
addpath(here);

ip = inputParser;
addParameter(ip, 'Fs',       128);
addParameter(ip, 'Duration', 4);
addParameter(ip, 'L',        48);
addParameter(ip, 'NumModes', 12);
addParameter(ip, 'Seed',     42);
addParameter(ip, 'SaveFig',  true);
addParameter(ip, 'OutDir',   fullfile(here, 'output'));
parse(ip, varargin{:});
p = ip.Results;

rng(p.Seed);
T = p.Fs * p.Duration;
t = (0:T-1)' / p.Fs;

fprintf('\n=== FED main demo ===\n');
fprintf('Fs=%d Hz, N=%d, L=%d, k=%d\n\n', p.Fs, T, p.L, p.NumModes);

%% Synthetic signal (Section 5.1 style)
c1 = 1.0 * sin(2*pi*3*t);
c2 = 0.7 * sin(2*pi*12*t) .* (0.6 + 0.5*sin(2*pi*0.7*t));
c3 = zeros(T, 1);
for bt = [0.5, 1.5, 2.5, 3.5]
    c3 = c3 + 0.5 * exp(-((t-bt).^2)/(2*0.1^2)) .* sin(2*pi*30*t);
end
x = c1 + c2 + c3 + 0.1 * randn(T, 1);

common = {'L', p.L, 'NumModes', p.NumModes, 'FStrategy', 'row', ...
    'Kmax', 6, 'Verbose', true};

fprintf('--- SSA baseline (F = I) ---\n');
out_ssa = fed_cov_main(x, common{:}, 'FTransform', 'identity', 'FParam', 1);

fprintf('\n--- FED default (F = HFD^2, p = 2) ---\n');
out_fed = fed_cov_main(x, common{:}, 'FTransform', 'power', 'FParam', 2);

fprintf('\n=== Summary ===\n');
fprintf('%-22s  epsL2   epsF_high   F-Sel   Finv_orth\n', 'Method');
fprintf('%s\n', repmat('-', 1, 58));
print_row('SSA (F=I)', out_ssa);
print_row('FED (HFD^2,p=2)', out_fed);

if p.SaveFig
    if ~isfolder(p.OutDir), mkdir(p.OutDir); end
    fig = figure('Color', 'w', 'Position', [100 100 1100 420], 'Visible', 'off');
    subplot(1, 2, 1);
    plot(t, x, 'Color', [0.2 0.2 0.2]); hold on;
    plot(t, out_ssa.x_hat, 'b', 'LineWidth', 1.1);
    grid on; xlabel('Time (s)'); ylabel('Amplitude');
    title(sprintf('SSA (F=I)  \\epsilon_{L2}=%.3f', out_ssa.epsilon_L2));
    legend('Signal', 'Reconstruction', 'Location', 'best');

    subplot(1, 2, 2);
    plot(t, x, 'Color', [0.2 0.2 0.2]); hold on;
    plot(t, out_fed.x_hat, 'Color', [0 0.45 0.74], 'LineWidth', 1.1);
    grid on; xlabel('Time (s)'); ylabel('Amplitude');
    title(sprintf('FED (HFD^2)  \\epsilon_{L2}=%.3f', out_fed.epsilon_L2));
    legend('Signal', 'Reconstruction', 'Location', 'best');

    fpath = fullfile(p.OutDir, 'demo_fed_main_reconstruction.png');
    if exist('exportgraphics', 'file') == 2
        exportgraphics(fig, fpath, 'Resolution', 150);
    else
        print(fig, fpath, '-dpng', '-r150');
    end
    close(fig);
    fprintf('\nFigure saved: %s\n', fpath);
end

fprintf('\nDone. Run verify_fed_main for the Step-6 algebra check.\n');
end

function print_row(name, o)
fprintf('%-22s  %5.3f   %9.3f   %5.3f   %8.1e\n', ...
    name, o.epsilon_L2, o.epsilon_F_local, o.F_selectivity, ...
    o.Finv_orthogonality_deviation);
end
